### PR TITLE
[lexical] Bug Fix: Prevent deleteLine from removing adjacent block decorators

### LINKED ISSUE #8722: Bug: DELETE_LINE_COMMAND on an empty ListItem removes the previous decorator
Lexical version: 0.45

## Steps To Reproduce

1. Insert the decorator node
2. Insert a list with an empty first listitem
3. Place the cursor on the first listitem and press Ctrl + Backspace

Link to code example: https://playground.lexical.dev/#doc=H4sIAAAAAAAAE9VWzW7bMAx-FYKXXbw4aTMs9XXFuj8MKzJgA4YdGImxtciSIVFJjaKXvcVOe8U9wmAHSdt06IAsKNaTQdL--JH8JPoSWRvxYSokjMUlBu-le6rKWB3YYfHlEqVtGAtsfZI0Y8xwySEa77AYZTj3oSbBAju_0exfn2KB-nw1Xp18Kj-rc7zKdvBuGF8z1Cawkh7OJWtvIxqn2QkWw2zDwpooRrjeobEkmxiL0Z1kmoWM7RE2wMMMa687MNd5LGYYpbW8Til80SV_UbFagE8CUjEorxm8A58C4JZL_-pNHgfLfmbkVZpB4MbHbkDtfUn3aaJb7DQwsN18KxRKlq1lpCO3NlKwWGAl0sQiz0sjVZoNlK_zOSmeeb_ILV8YRbaf-r7VDw5d7X2SOTrY0D5Yasvgk9NrvShyMGOY956HUE3FgR-bUHIJzHlNxuUNqQWVHDehp822oY9HTscHm-Ybbx7uwjk1UfmgYcphyeG_FJFeU-xVZNzSCOdv67Px6ty505N_EghQd2QrElgZqfoLX5jqB9XNeE_EHbTO9XEdnSVrWfqCKUgfFSqxwGTv7uR9W_eOotg2gxU_Cd2adFHIiW2BtDauBOW9BccrmDNJChxBPEhlIlwf7gFMPdS0YIgpMLQ-geq374zUArpLDVYVuz7g-EKgZAHqBuYU_-103FuY-AW7P9b16-eP79fIXPtvZqfTylKM76nexqGipmljbSz_T9dVQ4HKQE21w7_DfnnNqzOnGy57pOn_HG9RvcrQUpQpLVljMXo-GU2Oh5PhZHx0nGH0KSi-tTRvfo3DwfjZYIhXvwHxyqQ9ogoAAA

## The current behavior

The DecoratorNode before the first ListItem is removed but the ListItem itself being empty remains. The listitem is correctly removed if it is preceded by a regular element rather than a decorator

https://github.com/user-attachments/assets/afcbf120-9faa-4712-a357-76076086d8c1

## The expected behavior

The ListItem is removed, but the decorator remains

## Impact of fix

This is unexpected behavior since DELETE_LINE_COMMAND is expected to apply only to the current line when the decorator is on the previous line

There is another problem with DELETE_LINE_COMMAND when there is text in the ListItem, the command first deletes the text itself and only the second time deletes the item

### PR DESCRIPTION
**Pull request by mayrang**
## Description

`deleteLine` (Cmd+Backspace / Cmd+Delete) from an empty block removes an adjacent block-level decorator (HorizontalRule, PageBreak) when it should only delete the empty block. The `modify('extend', isBackward, 'lineboundary')` call extends the selection past block boundaries into the adjacent decorator, and the subsequent `removeText()` deletes the decorator along with the empty block.

After the lineboundary extend, check whether the resulting selection includes any non-inline `DecoratorNode`. If so, collapse back to the anchor and delegate to `deleteCharacter()`, which already handles decorator boundaries correctly — it removes the empty anchor block and selects the decorator via `NodeSelection`.

Closes #8722

## Test plan

- [x] `pnpm vitest run --project unit -t "DELETE_LINE_COMMAND"` — 6 tests pass:
  - `deleteCharacter` backward from empty ListItem preserves preceding decorator
  - `deleteLine` backward/forward from empty ListItem and empty paragraph preserves adjacent decorator (3 cases via `test.for`)
  - `deleteLine` backward from empty ListItem preserves preceding empty paragraph (regression guard)
  - `deleteLine` backward with preceding non-empty paragraph merges normally
- [x] `pnpm vitest run --project unit` — full suite passes (3337 tests).
- [x] `pnpm tsc --noEmit`, prettier / eslint clean.
- [x] Manual playground on macOS (Chrome, Safari, Firefox):
  - Insert horizontal rule → empty bullet list below → Cmd+Backspace: rule survives.
  - Empty paragraph after horizontal rule → Cmd+Backspace: rule survives.
  - Empty bullet list above horizontal rule → Cmd+Delete: rule survives.
  - Empty paragraph before empty bullet list → Cmd+Backspace: paragraph preserved (regression check).
  - Text before empty bullet list → Cmd+Backspace: normal merge.
  - Non-empty list item after horizontal rule → Cmd+Backspace: text deleted, rule untouched.
  - Empty editor → Cmd+Backspace: no change.

### CODE REVIEW

**Review by potatowagon — COMMENTED** [review]
Reviewed by Navi (Tater Thoughts Bobblehead) on behalf of @potatowagon.

**Assessment: LGTM ✅ — Ready to approve**

This is a well-crafted bug fix for #8722. The approach is sound and minimal.

**What I checked:**

1. **Logic correctness:** The fix correctly identifies when `deleteLine` overreaches past block-level decorators after the lineboundary extend. By collapsing the selection back to the anchor and delegating to `deleteCharacter`, it leverages the existing decorator-boundary logic rather than reimplementing it. This is the right abstraction layer to handle this.

2. **Edge cases covered by tests:**
   - Backward delete from empty ListItem with preceding decorator ✅
   - Backward delete from empty paragraph with preceding decorator ✅
   - Forward delete from empty ListItem with following decorator ✅
   - Backward delete should not remove preceding empty paragraph ✅
   - Backward delete from list item preceded by non-empty paragraph (normal merge behavior preserved) ✅

3. **No regressions:** The condition (`this.getNodes().some(node => $isDecoratorNode(node) && !node.isInline())`) is narrow — it only triggers when the extended selection actually contains a block-level decorator. Normal text deletion, inline decorator handling, and paragraph merging are unaffected.

4. **Placement in control flow:** The new branch sits between the "anchor === focus after extend" early return (which delegates to deleteCharacter) and the `removeText()` fallback. This is the correct insertion point — it catches the case where the extend DID move focus but overshot into a decorator.

5. **www compat:** No API changes, no export modifications, no changed function signatures. Purely behavioral fix in internal selection logic. Safe for internal consumers (MLCComposer, XDSRichText, EPS).

6. **CI:** All checks green — core-tests (unit Node 22/24, browser, integrity), e2e canary, CLA, Vercel deploys all pass.

**One minor observation:** The `.some()` call iterates all selected nodes on every deleteLine. For typical selections this is negligible, but in theory a very large selection could be slightly slower. Not a concern for real-world usage since deleteLine operates on single lines.

Ship it. 🚀

File: packages/lexical/src/LexicalSelection.ts [hunk 1]
```diff
@@ -2249,6 +2249,15 @@ export class RangeSelection implements BaseSelection {
...
+    ) {
+      // The lineboundary extend overreached past a block-level
+      // decorator (e.g. a page-break or horizontal-rule).
+      // Collapse back and let deleteCharacter handle the boundary
+      // — it already knows how to stop at decorators.
+      this.focus.set(this.anchor.key, this.anchor.offset, this.anchor.type);
+      this.deleteCharacter(isBackward);
     } else {
       this.removeText();
     }
```
Comment (etrepum) [inline packages/lexical/src/LexicalSelection.ts]: I think something like this would be a little more clear:
```ts
      const cr = $caretRangeFromSelection(this);
      if ($isDecoratorNode(cr.focus.origin) && !cr.focus.origin.isInline()) {
        $setPointFromCaret(this.focus, cr.anchor);
        this.deleteCharacter(isBackward);
      } else {
        this.removeText();
      }
```

The next problem is that similar strange behavior happens when the cursor moves into another block that isn't going to collapse properly with removeText(), e.g. if there's a TableNode or ReviewNode before the List -> empty ListItemNode.

It might just make sense to instead detect the beginning of line condition more directly (e.g. are we at the first cursor position in a block) and defer to deleteCharacter before trying to modify the selection.

Comment (mayrang) [reply packages/lexical/src/LexicalSelection.ts]: Switched to `$caretRangeFromSelection` + `$setPointFromCaret`, then generalized further: after the lineboundary extend, check whether focus crossed into a different block (`$findMatchingParent(node, INTERNAL_$isBlock)`). If it did, collapse back and defer to `deleteCharacter`. This handles decorators, tables, and any other non-collapsible block without per-type checks.

One test expected the preceding empty paragraph to survive a backward `deleteLine` from an empty list item — updated it, since in real browsers `modify` stays collapsed there and `deleteCharacter` handles the merge directly.

**Review by etrepum — APPROVED** [review]
